app/code/Magento/InventoryBundleIndexer/Indexer/SourceItem/SourceItemIndexer.php

```diff
@@ -0,0 +1,143 @@
+     */
+    private $bundleIndexDataProvider;
+
+    /**
+     * @var GetBundleChildrenSourceItemsIdsWithSku
+     */
+    private $getBundleChildrenSourceItemsIdsWithSku;
+
+    /**
+     * @var IndexNameBuilder
+     */
+    private $indexNameBuilder;
+
+    /**
+     * @var IndexHandlerInterface
+     */
+    private $indexHandler;
+
+    /**
+     * @param GetAllBundleChildrenSourceItemsIdsWithSku $getAllBundleChildrenSourceItemsIdsWithSku
+     * @param GetSkuListInStock $getSkuListInStock
+     * @param BundleIndexDataProvider $bundleIndexDataProvider
+     * @param GetBundleChildrenSourceItemsIdsWithSku $getBundleChildrenSourceItemsIdsWithSku
+     * @param IndexNameBuilder $indexNameBuilder
+     * @param IndexHandlerInterface $indexHandler
+     */
+    public function __construct(
+        GetAllBundleChildrenSourceItemsIdsWithSku $getAllBundleChildrenSourceItemsIdsWithSku,
+        GetSkuListInStock $getSkuListInStock,
+        BundleIndexDataProvider $bundleIndexDataProvider,
+        GetBundleChildrenSourceItemsIdsWithSku $getBundleChildrenSourceItemsIdsWithSku,
+        IndexNameBuilder $indexNameBuilder,
+        IndexHandlerInterface $indexHandler
+    ) {
+        $this->getAllBundleChildrenSourceItemsIdsWithSku = $getAllBundleChildrenSourceItemsIdsWithSku;
+        $this->getSkuListInStock = $getSkuListInStock;
+        $this->bundleIndexDataProvider = $bundleIndexDataProvider;
+        $this->getBundleChildrenSourceItemsIdsWithSku = $getBundleChildrenSourceItemsIdsWithSku;
+        $this->indexNameBuilder = $indexNameBuilder;
+        $this->indexHandler = $indexHandler;
+    }
+
+    /**
+     * @return void
+     */
+    public function executeFull()
+    {
+        $bundleChildrenSourceItemsIdsWithSku = $this->getAllBundleChildrenSourceItemsIdsWithSku->execute();
+
+        $this->executeByBundleChildrenSourceItemsIdsWithSku($bundleChildrenSourceItemsIdsWithSku);
+    }
+
+    /**
+     * @param int $sourceItemId
+     * @return void
+     */
+    public function executeRow(int $sourceItemId)
+    {
+        $this->executeList([$sourceItemId]);
+    }
+
+    /**
+     * @param array $sourceItemIds
+     * @return void
+     */
+    public function executeList(array $sourceItemIds)
+    {
+        $bundleChildrenSourceItemsIdsWithSku = $this->getBundleChildrenSourceItemsIdsWithSku->execute($sourceItemIds);
+
+        $this->executeByBundleChildrenSourceItemsIdsWithSku($bundleChildrenSourceItemsIdsWithSku);
+    }
+
+    /**
+     * @param array $bundleChildrenSourceItemsIdsWithSku
+     * @return void
+     */
+    private function executeByBundleChildrenSourceItemsIdsWithSku(array $bundleChildrenSourceItemsIdsWithSku)
+    {
+        foreach ($bundleChildrenSourceItemsIdsWithSku as $bundleSku => $bundleChildrenSourceItemsIds) {
+            $skuListInStockList = $this->getSkuListInStock->execute($bundleChildrenSourceItemsIds);
+            foreach ($skuListInStockList as $skuListInStock) {
```

Comment:
Can we load all data for all bundles in one query instead of having query in the cycle?

Reply:
Fixed.